src/main/src/test/java/org/geoserver/catalog/ResourcePoolTest.java

```diff
@@ -971,4 +980,290 @@ public void testCoverageReaderInputConverterFilePath() throws IOException {
         // File object
         assertTrue(reader.getSource() instanceof File);
     }
+
+    @Test
+    public void testGetDataStoreConcurrency() throws Exception {
+        ResourcePool pool = null;
+        try {
+            final Catalog catalog = getCatalog();
+            DataStoreInfo ds = storeInfo(catalog, "concurrencyTest1");
+            pool =
+                    new ResourcePool(getCatalog()) {
+                        @Override
+                        protected DataAccess<? extends FeatureType, ? extends Feature>
+                                createDataAccess(DataStoreInfo info, DataStoreInfo expandedStore)
+                                        throws IOException {
+                            return super.createDataAccess(info, expandedStore);
+                        }
+                    };
+
+            ResourcePoolLatchedThread.PoolBiFunction<DataStoreInfo, DataAccess> function =
+                    (resPool, info) -> resPool.getDataStore(info);
+            CountDownLatch taskLatch = new CountDownLatch(1);
+            int numberOfThreads = 5;
+            CountDownLatch completeLatch = new CountDownLatch(numberOfThreads);
+            List<ResourcePoolLatchedThread<DataStoreInfo, DataAccess>> threads =
+                    getLatchedThreads(
+                            taskLatch, completeLatch, numberOfThreads, pool, ds, function);
+            threads.forEach(t -> t.start());
```

Comment:
So you are explicitly creating threads and starting them... but I don't see you destroying them, meaning they will accumulate in VM. It's a test JVM, so no big deal, but could cause troubles in the future, as Maven surefire testing JVMs can be reused across modules. Best clean up and kill the threads, here and in the other tests. Or create runnables and submit them to a Executor (you'll still have to shut down the executor, anyways).